include/byte_order_generic_x86_64.h

```diff
@@ -17,65 +17,88 @@
   Optimized function-like macros for the x86 architecture (_WIN32 included).
 */
 
-#define sint2korr(A)	(int16) (*((int16 *) (A)))
-#define sint3korr(A)	((int32) ((((uchar) (A)[2]) & 128) ? \
-				  (((uint32) 255L << 24) | \
-				   (((uint32) (uchar) (A)[2]) << 16) |\
-				   (((uint32) (uchar) (A)[1]) << 8) | \
-				   ((uint32) (uchar) (A)[0])) : \
-				  (((uint32) (uchar) (A)[2]) << 16) |\
-				  (((uint32) (uchar) (A)[1]) << 8) | \
-				  ((uint32) (uchar) (A)[0])))
-#define sint4korr(A)	(int32)  (*((int32 *) (A)))
-#define uint2korr(A)	(uint16) (*((uint16 *) (A)))
-#define uint3korr(A)	(uint32) (((uint32) ((uchar) (A)[0])) +\
-				  (((uint32) ((uchar) (A)[1])) << 8) +\
-				  (((uint32) ((uchar) (A)[2])) << 16))
-#define uint4korr(A)	(uint32) (*((uint32 *) (A)))
-
-
-static inline ulonglong uint5korr(const void *p)
+#include <string.h>
+
+inline static int16 sint2korr(const void *p)
+{
+  int16 result= 0;
+  memcpy(&result, p, 2);
+  return result;
+}
+
+inline static int32 sint3korr(const void *p)
+{
+  int32 result= 0;
+  const char *ptr= (const char *) p;
+  result= ptr[2];
+  result<<= 16;
+  memcpy(&result, ptr, 2);
```

Comment:
Why not just `memcpy(&result, ptr, 3);`?

Reply:
Because the last byte is `0` for positives and `256` for negatives here. Original code even have a branch to check that so it's a small performance benefit here.

Reply:
Then why sint2korr() is different?

Reply:
Because with `sint2korr()` I read 2 bytes in `int16` and initialize every byte of it.
With `sint3korr()` I just cant initialize 3 bytes of `int32`. I need to will it with `0` or `256`.

Reply:
Sorry, overlooked that you memcpy() into int16 there.